### PR TITLE
fix: create typeorm_metatable when running migrations

### PR DESCRIPTION
**Pull request by BitPatty**
Proposed fix for #4923 
When generating migrations for views a `typeorm_metadata` table is created in which the views select statements are stored during the migration. However, the table won't be created during the actual execution of the migration. Thus, running a migration on a database which doesn't have the `typeorm_metadata` table will eventually fail. To avoid that, the table is created during a migration
if necessary.

### CODE REVIEW

**Comment by pablosproject** [comment]
Any chances we got this merged any soon? Just ran into this issue and I think is important to have this in the library.

File: src/schema-builder/RdbmsSchemaBuilder.ts [hunk 1]
```diff
@@ -98,17 +95,24 @@ export class RdbmsSchemaBuilder implements SchemaBuilder {
         }
     }
 
+    /**
+     * If the schema contains views, create the typeorm_metadata table if it doesn't exist yet
+     */
+    async createMetadataTableIfNecessary(): Promise<void> {
+        if (!this.queryRunner) this.queryRunner = this.connection.createQueryRunner("master");
```
Comment (pleerock) [inline src/schema-builder/RdbmsSchemaBuilder.ts]: why did you add this check? it looks like this.queryRunner is created in all places where we use this method. Having it just for type is super confusing. But if its really needed (but I don't think so) it should also be properly released.

File: src/commands/MigrationRunCommand.ts [hunk 2]
```diff
@@ -73,6 +74,12 @@ export class MigrationRunCommand implements yargs.CommandModule {
                     // noop
             }
 
+            const schemaBuilder = connection.driver.createSchemaBuilder();
+
+            if (schemaBuilder instanceof RdbmsSchemaBuilder) {
+                await schemaBuilder.createMetadataTableIfNecessary();
+            }
```
Comment (pleerock) [inline src/commands/MigrationRunCommand.ts]: can we make same inside MigrationExecutor#executePendingMigrations ? Near "await this.createMigrationsTableIfNotExist(queryRunner);" line to make sure its consistent and works not only when called from the CLI

Comment (pleerock) [reply src/commands/MigrationRunCommand.ts]: I guess need to remove this code now?

**Comment by GSKryachko** [comment]
I also would like to see it merged as soon as possible.

**Comment by pleerock** [comment]
This PR isn't finished. For some reason @BitPatty added metadata table creation code into `undoLastMigration` method, I'm completely confused why

**Comment by BitPatty** [comment]
I'm just as confused on how this happened :/

**Comment by michaeldever** [comment]
This would be awesome, hitting my head against this in test for the last little while.

**Comment by coyoteecd** [comment]
@pleerock can you merge and release this? I've run into this too

**Comment by imnotjames** [comment]
Can you fix the conflict

**Comment by pleerock** [comment]
we don't have tests for cli commands yet.. if someone can contribute and start writing tests for CLI for the first time, this one is a good opportunity. If nobody wants, probably we can merge this one.

**Comment by BitPatty** [comment]
@imnotjames fixed the conflict

**Comment by BitPatty** [comment]
@imnotjames Can you check the tests?

**Comment by imnotjames** [comment]
> @imnotjames Can you check the tests?

They're currently failing, it looks like?

**Comment by BitPatty** [comment]
The CI build had a timeout during service setup, it didn't run the tests

**Comment by pleerock** [comment]
@AlexMesser can you please review this PR?

**Comment by pleerock** [comment]
Okay, I have to review it by myself.

File: src/schema-builder/RdbmsSchemaBuilder.ts [hunk 3]
```diff
@@ -99,17 +96,24 @@ export class RdbmsSchemaBuilder implements SchemaBuilder {
         }
     }
 
+    /**
+     * If the schema contains views, create the typeorm_metadata table if it doesn't exist yet
+     */
+    async createMetadataTableIfNecessary(): Promise<void> {
+        if (this.viewEntityToSyncMetadatas.length > 0) {
+            this.queryRunner = this.queryRunner || this.connection.createQueryRunner();
```
Comment (pleerock) [inline src/schema-builder/RdbmsSchemaBuilder.ts:104]: `|| this.connection.createQueryRunner()` this code can be very dangerous since its not clear here where it gets **released**. Use `this.queryRunner` if it's available, or pass it to `createMetadataTableIfNecessary` method from the places where you call this method.

Comment (BitPatty) [reply src/schema-builder/RdbmsSchemaBuilder.ts:104]: On a side note, with both the current implementation [1]/[2] as well as this PR the migration table is created when **generating** migrations. Is this intended behavior?

[1] https://github.com/typeorm/typeorm/blob/a92120ac8821461055993a36ddba17d3466de925/src/commands/MigrationGenerateCommand.ts#L105
[2] https://github.com/typeorm/typeorm/blob/a92120ac8821461055993a36ddba17d3466de925/src/schema-builder/RdbmsSchemaBuilder.ts#L116-L125

Comment (pleerock) [reply src/schema-builder/RdbmsSchemaBuilder.ts:104]: > the migration table is created when generating migrations. Is this intended behavior?

Probably it shouldn't be inside `log` method, since you are creating metadata table in `executePendingMigrations`, right?

**Comment by TheNoim** [comment]
I just noticed this could also be relevant for my pr #6469, because it uses the ´typeorm_metadata´ table.

**Comment by AlexMesser** [comment]
thank you for contribution!

**Comment by issackr** [comment]
@pleerock @BitPatty 
Hi
I am getting this error while having my server that run migrations going up
typeorm_metadata table already exists
It looks like this behaviour started around the merging of this fix, do you think it might be connected?

Thanks

**Comment by BitPatty** [comment]
You might have patched a migration manually to explicitly create the typeorm_metadata table. Check your migration files for any `CREATE TABLE typeorm_metadata` statements.
